tests/integration/targets/firewall_resource/tasks/test.yml

```diff
@@ -0,0 +1,95 @@
+# Copyright: (c) 2020, Hetzner Cloud GmbH <info@hetzner-cloud.de>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+---
+- name: Test missing required parameters
+  hetzner.hcloud.firewall_resource:
+    firewall: "{{ hcloud_firewall_name }}"
+    state: present
+  ignore_errors: true
+  register: result
+- name: Verify missing required parameters
+  ansible.builtin.assert:
+    that:
+      - result is failed
+      - 'result.msg == "one of the following is required: servers, label_selectors"'
+
+- name: Test create with check mode
+  hetzner.hcloud.firewall_resource:
+    firewall: "{{ hcloud_firewall_name }}"
+    servers: ["{{ hcloud_server_name }}"]
+  check_mode: true
+  register: result
+- name: Verify create with check mode
+  ansible.builtin.assert:
+    that:
+      - result is changed
+
+- name: Test create
+  hetzner.hcloud.firewall_resource:
+    firewall: "{{ hcloud_firewall_name }}"
+    servers: ["{{ hcloud_server_name }}"]
+  register: result
+- name: Verify create
+  ansible.builtin.assert:
+    that:
+      - result is changed
+      - result.hcloud_firewall_resource.firewall == hcloud_firewall_name
+      - result.hcloud_firewall_resource.servers | list | count == 1
+      - result.hcloud_firewall_resource.servers[0] == hcloud_server_name
+      - result.hcloud_firewall_resource.label_selectors | list | count == 0
+
+- name: Test create idempotency
+  hetzner.hcloud.firewall_resource:
+    firewall: "{{ hcloud_firewall_name }}"
+    servers: ["{{ hcloud_server_name }}"]
+  register: result
+- name: Verify create idempotency
+  ansible.builtin.assert:
+    that:
+      - result is not changed
+
+- name: Test update
```

Comment:
Missing: the result of an update to a label_selector should remove the server attribute from the firewall (I guess)?

Reply:
The current behavior of the ansible module matches the API. You can add and remove resources independently.

(Not an ansible expert) This seems to match the general Ansible behavior, only the fields you specify are actually updated, there is no expectation that you specify the full configuration in every module invocation. Ie. you do not specify all installed OS packages when you want to install a new one.

I agree that magically removing the other attribute would not match my expectation.

That said, is there a way to remove all current attachements/resources from a firewall? Probably through the `hcloud_firewall_info` resource and `state: absent`, right?

Reply:
There is no easy way to remove all the attached resources, using  `hcloud_firewall_info` is an option, but it is indeed suboptimal.

Maybe having an extra firewall_resource state `clean` could be the solution to this?

Reply:
I prefer to implement such feature in a future PR, this one is already quite large.